presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushAggregatesIntoTableScanRuleSet.java

```diff
@@ -0,0 +1,190 @@
+        public Pattern<AggregationNode> getPattern()
+        {
+            return AGG_WITHOUT_PROJECTION;
+        }
+    }
+
+    private abstract class BasePushAggregationToTableScan
+            implements Rule<AggregationNode>
+    {
+        @Override
+        public boolean isEnabled(Session session)
+        {
+            return isEnableAggPushdown(session);
+        }
+
+        @Override
+        public Result apply(AggregationNode node, Captures captures, Context context)
+        {
+            TableScanNode tableScan = captures.get(TABLE_SCAN);
+            Map<String, ColumnHandle> assignments = tableScan.getAssignments()
+                    .entrySet().stream()
+                    .collect(toImmutableMap(entry -> entry.getKey().getName(), Map.Entry::getValue));
+            Map<Symbol, AggregationNode.Aggregation> aggregations = node.getAggregations();
+
+            ImmutableList<AggregateFunction> aggregateFunctions = toAggFunction(aggregations);
+
+            if (aggregateFunctions.isEmpty()) {
+                return Result.empty();
+            }
+
+            GroupingSetDescriptor groupingSets = node.getGroupingSets();
+
+            if (groupingSets.getGroupingSetCount() > 1) {
+                // Let the PushAggregationToTableScanWithGroupingSets rule handle this.
+                return Result.empty();
+            }
+
+            List<ColumnHandle> groupByColumns = groupingSets.getGroupingKeys().stream()
+                    .map(assignments::get)
+                    .collect(toList());
+
+            Optional<AggregatePushdownResult<TableHandle>> connectorTableHandleAggregatePushdownResult = metadata.applyAggregation(context.getSession(),
+                    tableScan.getTable(),
+                    aggregateFunctions,
+                    assignments,
+                    Optional.of(ImmutableList.of(groupByColumns)));
+
+            if (!connectorTableHandleAggregatePushdownResult.isPresent()) {
+                return Result.empty();
+            }
+
+            AggregatePushdownResult<TableHandle> aggPushdownResult = connectorTableHandleAggregatePushdownResult.get();
+
+            TableHandle handle = aggPushdownResult.getHandle();
+            ImmutableList.Builder<Symbol> newScanOutputs = new ImmutableList.Builder<>();
+            ImmutableMap.Builder<Symbol, ColumnHandle> newScanAssignments = new ImmutableMap.Builder<>();
+            Map<String, ProjectionApplicationResult.Assignment> nameToAssignment = aggPushdownResult.getAssignments().stream()
+                    .collect(Collectors.toMap(ProjectionApplicationResult.Assignment::getVariable, Function.identity()));
+            for (Symbol aggOutputSymbol : node.getOutputSymbols()) {
+                ProjectionApplicationResult.Assignment assignment = nameToAssignment.get(aggOutputSymbol.getName());
+
+                if (assignment == null) {
+                    //TODO We pushdown all or nothing for V1 but we need to extend this in future.
```

Comment:
I don't see how we could do partial pushdown. Replace TODO comment with something objective, but implying a change in the future (a change may or may not happen)

Also, isn't it an error condition (throw)?

Reply:
I wasn't considering this as an error condition as I wanted to let the connector return a partial set and let the rule just ignore it for now and whenever we figure out how to do partial aggregations we could remove the checks from here. I have removed the comment and the verify method now throws the exception.

I had some ideas about how partial aggs (when connector can support a sub set of aggregate functions completely and not when connectors are returning partially aggregated results) would work at least in some of the cases but overall design needs more thinking.